File: 01-week-1--starter-algorithms/04-day-5--find-shortest-string/javascript/find_shortest_string.js
function findShortestString(arr) {
  // type your code here
  let answer = (!!arr[0] ? arr[0] : "")
  for (let i = 0; i < arr.length; i++) {
    if (arr[i].length < answer.length) {answer = arr[i]}
    if (i === arr.length - 1) {return answer}
  }
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 'a'");
  console.log("=>", findShortestString(['aaa', 'a', 'bb', 'ccc']));

  console.log("");

  console.log("Expecting: 'hi'");
  console.log("=>", findShortestString(['cat', 'hi', 'dog', 'an']));

  console.log("");

  console.log("Expecting: 'lily'");
  console.log("=>", findShortestString(['flower', 'juniper', 'lily', 'dadelion']));

  // BENCHMARK HERE
}

module.exports = findShortestString;

// Please add your pseudocode to this file
/*
  -create variable *answer equal to an empty string
  -iterate through arr, test element string length, if element.length < answer.length, answer = element
   if element === arr.last return answer
*/
// And a written explanation of your solution
/*
  I am creating a variable *answer set to the first element in arr or a blank string and testing its length against each
  element in the argument array.  If an elements length is less than answer I am assigning the value of answer
  to that element.  If the current element in the iteration is the last one of the array, I'm returning answer
*/

File: 01-week-1--starter-algorithms/04-day-5--find-shortest-string/ruby/find_shortest_string.rb
def find_shortest_string(arr)
  # type your code in here
  answer = (!!arr[0] ? arr[0] : "")
  arr.each do |element|
    if element.length < answer.length
      answer = element
    end
  end
  answer
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 'a'"
  puts "=>", find_shortest_string(['aaa', 'a', 'bb', 'ccc'])

  puts

  puts "Expecting: 'hi'"
  puts "=>", find_shortest_string(['cat', 'hi', 'dog', 'an'])

  puts

  puts "Expecting: 'lily'"
  puts "=>", find_shortest_string(['flower', 'juniper', 'lily', 'dadelion'])

  # Don't forget to add your own!

  # BENCHMARK HERE
end

# Please add your pseudocode to this file
=begin
  -create variable *answer equal to the first element in arr or an empty string
  -iterate through arr, test element string length, if element.length < answer.length, answer = element
  -return answer
=end
# And a written explanation of your solution
=begin
  I am creating a variable *answer set to the first element in arr or a blank string and testing its length against each
  element in the argument array.  If an elements length is greater answer I am assigning the blank steam variable
  to that element.  After this iteration is complete I am returning answer
=end